File: src/go/cdk/productPricingService/productPricingService.go
//
// Unless explicitly stated otherwise all files in this repository are licensed
// under the Apache License Version 2.0.
// This product includes software developed at Datadog (https://www.datadoghq.com/).
// Copyright 2024 Datadog, Inc.
//

package productpricingservice

import (
	sharedprops "cdk/shared"
	sharedconstructs "cdk/sharedConstructs"
	"fmt"

	"github.com/aws/aws-cdk-go/awscdk/v2/awslambdaeventsources"
	"github.com/aws/aws-cdk-go/awscdk/v2/awssns"
	"github.com/aws/aws-cdk-go/awscdk/v2/awssqs"
	"github.com/aws/aws-cdk-go/awscdk/v2/awsssm"
	"github.com/aws/constructs-go/constructs/v10"
	"github.com/aws/jsii-runtime-go"
)

type ProductPricingServiceProps struct {
	SharedProps         sharedprops.SharedProps
	ProductCreatedTopic awssns.ITopic
	ProductUpdatedTopic awssns.ITopic
}

func NewProductPricingService(scope constructs.Construct, id string, props *ProductPricingServiceProps) {
	priceCalculatedTopic := awssns.NewTopic(scope, jsii.String("GoPriceCalculatedTopic"), &awssns.TopicProps{})

	environmentVariables := make(map[string]*string)
	environmentVariables["ENV"] = jsii.String(props.SharedProps.Env)
	environmentVariables["PRICE_CALCULATED_TOPIC_ARN"] = jsii.String(*priceCalculatedTopic.TopicArn())

	handleProductCreatedFunction := sharedconstructs.NewInstrumentedFunction(scope, "PricingHandleProductCreated", &sharedconstructs.InstrumentedFunctionProps{
		SharedProps:          props.SharedProps,
		Entry:                "../src/product-pricing-service/handle-product-created/",
		FunctionName:         fmt.Sprintf("GoPricingHandlerProductCreated-%s", props.SharedProps.Env),
		EnvironmentVariables: environmentVariables,
	})

	priceCalculatedTopic.GrantPublish(handleProductCreatedFunction.Function)

	handleProductCreatedDLQ := awssqs.NewQueue(scope, jsii.String("GoPricingProductCreatedDLQ"), &awssqs.QueueProps{
		QueueName: jsii.Sprintf("GoPricingHandleCreatedDLQ-%s", props.SharedProps.Env),
	})
	handleProductCreatedFunction.Function.AddEventSource(awslambdaeventsources.NewSnsEventSource(props.ProductCreatedTopic, &awslambdaeventsources.SnsEventSourceProps{
		DeadLetterQueue: handleProductCreatedDLQ,
	}))

	handleProductUpdatedFunction := sharedconstructs.NewInstrumentedFunction(scope, "PricingHandleProductUpdated", &sharedconstructs.InstrumentedFunctionProps{
		SharedProps:          props.SharedProps,
		Entry:                "../src/product-pricing-service/handle-product-updated/",
		FunctionName:         fmt.Sprintf("GoPricingHandlerProductUpdated-%s", props.SharedProps.Env),
		EnvironmentVariables: environmentVariables,
	})
	priceCalculatedTopic.GrantPublish(handleProductUpdatedFunction.Function)

	handleProductUpdatedDLQ := awssqs.NewQueue(scope, jsii.String("GoPricingProductUpdatedDLQ"), &awssqs.QueueProps{
		QueueName: jsii.Sprintf("GoPricingHandleUpdatedDLQ-%s", props.SharedProps.Env),
	})
	handleProductUpdatedFunction.Function.AddEventSource(awslambdaeventsources.NewSnsEventSource(props.ProductUpdatedTopic, &awslambdaeventsources.SnsEventSourceProps{
		DeadLetterQueue: handleProductUpdatedDLQ,
	}))

	awsssm.NewStringParameter(scope, jsii.String("GoProductCreatedTopicArn"), &awsssm.StringParameterProps{
		ParameterName: jsii.String("/go/product/pricing-calculated-topic"),
		StringValue:   priceCalculatedTopic.TopicArn(),
	})

	awsssm.NewStringParameter(scope, jsii.String("GoProductCreatedTopicName"), &awsssm.StringParameterProps{
		ParameterName: jsii.String("/go/product/pricing-calculated-topic-name"),
		StringValue:   priceCalculatedTopic.TopicName(),
	})
}

File: src/order-service/cdk/Services/Product.Pricing/ProductPricingService.cs
// Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
// This product includes software developed at Datadog (https://www.datadoghq.com/).
// Copyright 2024 Datadog, Inc.

using System.Collections.Generic;
using Amazon.CDK.AWS.Lambda.EventSources;
using Amazon.CDK.AWS.SecretsManager;
using Amazon.CDK.AWS.SNS;
using Amazon.CDK.AWS.SSM;
using Constructs;
using ServerlessGettingStarted.CDK.Constructs;

namespace ServerlessGettingStarted.CDK.Services.Product.Pricing;

public record ProductPricingServiceProps(SharedProps Shared, ISecret DdApiKeySecret, ITopic ProductCreatedTopic, ITopic ProductUpdatedTopic);

public class ProductPricingService : Construct
{
    public ProductPricingService(Construct scope, string id, ProductPricingServiceProps props) : base(scope, id)
    {
        var productPricingUpdatedTopic = new Topic(this, "ProductPricingUpdated", new TopicProps()
        {
            TopicName = $"DotnetProductPricingUpdated-{props.Shared.Env}"
        });
        
        var apiEnvironmentVariables = new Dictionary<string, string>(2)
        {
            { "PRICE_CALCULATED_TOPIC_ARN", productPricingUpdatedTopic.TopicArn },
        };
        
        var handleProductCreatedFunction = new InstrumentedFunction(this, "HandleProductCreatedFunction",
            new FunctionProps(props.Shared,"HandleProductCreated", "../src/Product.Pricing/ProductPricingService.Lambda/",
                "ProductPricingService.Lambda::ProductPricingService.Lambda.Functions_HandleProductCreated_Generated::HandleProductCreated", apiEnvironmentVariables, props.DdApiKeySecret));
        handleProductCreatedFunction.Function.AddEventSource(new SnsEventSource(props.ProductCreatedTopic));
        productPricingUpdatedTopic.GrantPublish(handleProductCreatedFunction.Function);
        
        var handleProductUpdatedFunction = new InstrumentedFunction(this, "HandleProductUpdatedFunction",
            new FunctionProps(props.Shared,"HandleProductUpdated", "../src/Product.Pricing/ProductPricingService.Lambda/",
                "ProductPricingService.Lambda::ProductPricingService.Lambda.Functions_HandleProductUpdated_Generated::HandleProductUpdated", apiEnvironmentVariables, props.DdApiKeySecret));
        handleProductUpdatedFunction.Function.AddEventSource(new SnsEventSource(props.ProductUpdatedTopic));
        productPricingUpdatedTopic.GrantPublish(handleProductUpdatedFunction.Function);
        
        var pricingUpdatedTopic = new StringParameter(this, "PricingUpdatedTopicArn",
            new StringParameterProps()
            {
                ParameterName = "/dotnet/product-pricing/pricing-updated-topic",
                StringValue = productPricingUpdatedTopic.TopicArn
            });
    }
}